src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs

```diff
@@ -663,58 +694,83 @@ private BoundExpression MakeImplicitConversion(BoundExpression rewrittenOperand,
-            }
+            return nullableNeverHasValue(expression);
 
-            // new int?() never has a value if there is no argument.
-            if (expression.Kind == BoundKind.ObjectCreationExpression)
+            bool nullableNeverHasValue(BoundExpression expression)
             {
-                if (((BoundObjectCreationExpression)expression).Arguments.Length == 0)
+                Debug.Assert(expression.Type.IsNullableType());
+                switch (expression)
                 {
-                    return true;
+                    case { ConstantValue: { Discriminator: ConstantValueTypeDiscriminator.Null } }:
+                    case BoundConversion { Conversion: { Kind: ConversionKind.DefaultOrNullLiteral } }:
+                    case BoundObjectCreationExpression { Arguments: { Length: 0 } }: // new int?() never has a value if there is no argument.
+                    case BoundDefaultExpression _:
+                        return true;
+                    case BoundConversion { Conversion: { IsNullable: true }, Operand: var operand } when operand.Type.IsNullableType():
+                        return nullableNeverHasValue(operand);
+                    default:
+                        return false;
                 }
             }
-
-            return false;
         }
 
         /// <summary>
         /// If the nullable expression always has a value, returns the value, otherwise null.
+        /// This is normally performed on a lowered expression, however for the purpose of
+        /// tuples and tuple equality operators, we do this on a partially lowered expression in
+        /// which conversions appearing at the top of the expression have not been lowered.
         /// If this method is updated to recognize more complex patterns, callers should be reviewed.
         /// </summary>
         private static BoundExpression NullableAlwaysHasValue(BoundExpression expression)
         {
-            if (expression.Type.IsNullableType() && expression.Kind == BoundKind.ObjectCreationExpression)
+            if (!expression.Type.IsNullableType())
+                return null;
+
+            switch (expression)
             {
-                BoundObjectCreationExpression objectCreation = (BoundObjectCreationExpression)expression;
-                if (objectCreation.Arguments.Length == 1)
-                {
-                    return objectCreation.Arguments[0];
-                }
+                // Detect the lowered nullable conversion from value type K to type Nullable<K>
+                case BoundObjectCreationExpression { Arguments: { Length: 1 } args }:
+                    return args[0];
+
+                // Detect the unlowered nullable conversion from value type K to type Nullable<K>
+                // This arises in lowering tuple equality operators
+                case BoundConversion
+                { Conversion: { Kind: ConversionKind.ImplicitNullable }, Operand: var convertedArgument }
```

Comment:
I'm really not a fan of this formatting, as this is where the opening brace would go if there was a `:` on the previous line. Can we indent this, or reformat on onto the previous line? Same for the next case.